In [44]:
# Import all the necessary packages.
import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.stats.api as sms
import sklearn
import matplotlib.pyplot as plt

from sklearn import linear_model
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LinearRegression
from statsmodels.formula.api import ols

# Note: Indicates situations that aren’t necessarily exceptions.
import warnings  
warnings.filterwarnings('ignore')  

In [2]:
# Import the salary_data.csv file.
data = pd.read_csv('ecommerce_data.csv') 

# Print the DataFrame.
(data.head())

Sale  por_OS  por_NON  recc  avg_no_it   age     dis  diff_reg  tax  \
0  0.63    18.0     2.31     0      6.575  65.2  4.0900         1  296   
1  2.73     0.0     7.07     0      6.421  78.9  4.9671         2  242   
2  2.73     0.0     7.07     0      7.185  61.1  4.9671         2  242   
3  3.24     0.0     2.18     0      6.998  45.8  6.0622         3  222   
4  6.91     0.0     2.18     0      7.147  54.2  6.0622         3  222   

       bk  lowstat  Median_s  
0  396.90     4.98      24.0  
1  396.90     9.14      21.6  
2  392.83     4.03      34.7  
3  394.63     2.94      33.4  
4  396.90     5.33      36.2

## 3 Define variables

In [3]:
data2 = data[['avg_no_it', 'tax', 'age', 'Median_s']]
data2.head()

avg_no_it  tax   age  Median_s
0      6.575  296  65.2      24.0
1      6.421  242  78.9      21.6
2      7.185  242  61.1      34.7
3      6.998  222  45.8      33.4
4      7.147  222  54.2      36.2

Dependent variable = Median_S

In [20]:
# Define the dependent variable.
y = data2['Median_s']  

# Define the independent variable.
X = data2[['avg_no_it', 'tax']]

In [21]:
# Fit the regression model.
mlr = linear_model.LinearRegression()
mlr.fit(X, y)

LinearRegression()

In [22]:
# Call the predictions for X (array).
mlr.predict(X)

array([26.63108645, 26.25540212, 32.36181073, 31.18391582, 32.37482535,
       26.64407278, 21.89365442, 23.17248345, 18.8484428 , 21.82971297,
       24.81098314, 21.86967638, 20.91055461, 21.4534628 , 22.62838696,
       20.53430443, 21.34156526, 21.78116273, 17.51307086, 19.67908752,
       18.42423654, 21.5813457 , 22.99605031, 20.36645812, 21.25364576,
       18.6560243 , 20.36645812, 22.23674558, 25.81746685, 27.24815682,
       19.56718998, 22.43656261, 21.46145548, 19.4712778 , 22.62838696,
       21.76901102, 21.03368433, 21.10561847, 22.03276951, 27.48776043,
       30.91662076, 29.18737937, 24.38377784, 24.71947046, 23.5845097 ,
       20.49134199, 21.32258085, 23.27279512, 18.22941315, 19.85192747,
       22.57891721, 23.79380478, 26.95890662, 22.85866105, 18.4003434 ,
       33.12673155, 24.82726558, 29.19079572, 23.38427535, 21.6418708 ,
       20.15523206, 21.95358538, 25.86999928, 28.31575979, 32.12616101,
       23.70386238, 19.68354362, 20.28418302, 18.0142615 , 20.34

In [23]:
# Print the R-squared value.
print("R-squared: ", mlr.score(X,y))  

# Print the intercept.
print("Intercept: ", mlr.intercept_) 

# Print the coefficients.
print("Coefficients:")  

# Map a similar index of multiple containers (to be used as a single entity).
list(zip(X, mlr.coef_))  

R-squared:  0.5605639377690896
Intercept:  -21.23309336056217
Coefficients:


[('avg_no_it', 7.992681419323308), ('tax', -0.015836826081673555)]

In [24]:
data2.describe()

avg_no_it         tax         age    Median_s
count  506.000000  506.000000  506.000000  506.000000
mean     6.284634  408.237154   68.574901   22.532806
std      0.702617  168.537116   28.148861    9.197104
min      3.561000  187.000000    2.900000    5.000000
25%      5.885500  279.000000   45.025000   17.025000
50%      6.208500  330.000000   77.500000   21.200000
75%      6.623500  666.000000   94.075000   25.000000
max      8.780000  711.000000  100.000000   50.000000

## 5. make predictions

In [25]:
# Create 'New_Distance' and define it as 15.2.
New_avg_no_it = 7.0 

# Create 'New_Distance' and define it as 15.2.
New_tax = 7.0 

# Print the predicted value. 
print ("Predicted Value: \n", mlr.predict([[New_avg_no_it, New_tax]]))  

Predicted Value: 
 [34.60481879]


## Train and test the MLR

In [26]:
# Split the data in 'train' (80%) and 'test' (20%) sets.
X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, y,
                                                                            test_size = 0.20,
                                                                            random_state = 42)

In [27]:
# Training the model using the 'statsmodel' OLS library.
# Fit the model with the added constant.
model = sm.OLS(Y_train, sm.add_constant(X_train)).fit()

# Set the predicted response vector.
Y_pred = model.predict(sm.add_constant(X_test)) 

# Call a summary of the model.
print_model = model.summary()

# Print the summary.
print(print_model)  

                            OLS Regression Results                            
Dep. Variable:               Median_s   R-squared:                       0.578
Model:                            OLS   Adj. R-squared:                  0.576
Method:                 Least Squares   F-statistic:                     275.1
Date:                Sat, 03 Jun 2023   Prob (F-statistic):           6.00e-76
Time:                        18:10:12   Log-Likelihood:                -1300.6
No. Observations:                 404   AIC:                             2607.
Df Residuals:                     401   BIC:                             2619.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -23.2468      3.125     -7.438      0.0

In [28]:
print(mlr.score(X_train, Y_train))

0.577871370129925


## Run a regression test

In [30]:
# Specify the model.
mlr = LinearRegression()  

# Fit the model. We can only fit the model with the training data set.
mlr.fit(X_train, Y_train)  


LinearRegression()

In [40]:
# Call the predictions for X in the train set.
y_pred_mlr = mlr.predict(X_test)  

# Print the predictions.
print("Prediction for test set: {}".format(y_pred_mlr))

Prediction for test set: [25.34068391 28.84665445 15.16523225 22.28317838 18.48624604 22.58581768
 21.0450819  21.0955331  17.92155281 20.60411364 21.79627197 21.68148737
  3.72117897 22.195515   19.60392283 24.15809126 22.5935717   6.71903567
 38.1067165  17.55616307 25.03651525 26.55466429 19.32835117 27.17520438
 16.06969119 11.26149444 19.22257124 22.06540459 21.85515612 18.6692089
 18.33184443 25.56641898 26.92342722 10.72171415 16.04477826 17.77308642
 32.58285169 21.93976077 22.71880459 23.77795457 15.31426677 30.10969603
 39.91859664 17.73766088 25.76120942 13.63652772 19.22869942 25.20629627
 17.40668545 26.66106955 21.72300893 31.19303715 22.89090109 26.97654765
 34.60077812 18.61081072 17.83850969 29.84469056 25.89334589 18.06682824
 28.40423186 33.36800226 26.29566909 17.19414294 24.87100073 16.73969981
 19.93119636 24.63181224 27.52476089 15.82056183 21.1078109  22.24809946
 14.80743574 20.75213693 21.43151828 11.47740655 21.87769997 38.41551614
 20.54571546 10.44007862 20

In [37]:
print(mlr.score(X_test, Y_test))

0.464835937616622


## Check for multicollinearity

In [39]:
# Check multicollinearity.
x_temp = sm.add_constant(X_train)

# Create an empty DataFrame. 
vif = pd.DataFrame()

# Calculate the VIF for each value.
vif['VIF Factor'] = [variance_inflation_factor(x_temp.values,
                                               i) for i in range(x_temp.values.shape[1])]

# Create the feature columns.
vif['features'] = x_temp.columns

# Print the values to one decimal points.
print(vif.round(1))

   VIF Factor   features
0       107.0      const
1         1.1  avg_no_it
2         1.1        tax


## Understanding the variance inflation value
the weaker the correlation between the IVs, the smaller the VIF will be. The best case scenario for the VIF will have a value of 1.0. The closer to 1.0, the happier we are, and the lower the risk of multicollinearity

In [45]:
# Determine heteroscedasticity.
model = sms.het_breuschpagan(model.resid, model.model.exog) 

In [46]:
terms = ['LM stat', 'LM Test p-value', 'F-stat', 'F-test p-value']
print(dict(zip(terms, model)))

{'LM stat': 24.21765359428799, 'LM Test p-value': 5.510656487586983e-06, 'F-stat': 12.785321886624978, 'F-test p-value': 4.142460297790512e-06}


We need the p-value to determine homoscedasticity / heteroscedasticity. If the calculated p-value is greater than 0.05, we fail to reject the H0 (null hypothesis), and assume that homoscedasticity is present. Since the LM-test p-value is 5.51..., we assume homoscedasticity.

In [48]:
# Call the metrics.mean_absolute_error function.  
print('Mean Absolute Error (Final):', metrics.mean_absolute_error(Y_test, Y_pred))  

# Call the metrics.mean_squared_error function.
print('Mean Square Error (Final):', metrics.mean_squared_error(Y_test, Y_pred))  

Mean Absolute Error (Final): 3.8350138001428684
Mean Square Error (Final): 39.24560530602035


Mean aboslute error - average difference between actual values and predicted values. average prediction errors. The closer the MAE to 0, the better. 
Mean square error - an alternative to the MAE, a compliment. Average square of the differences between original values and predicted values. Squaring means all values become positive.